### ISSUE TITLE
Gherkin Scope Issue?

### ISSUE THREAD

**Issue by wekempf** [pester/Pester#1071]
## 1. Provide a general summary of the issue in the Title above

I've got some Gherkin tests that work for me locally, but fail in my AppVeyor builds, and I don't know why. I know PowerShell moderately well, and I'm familiar with Gherkin, so when I found out Pester had Gherkin support I decided to use it. Unfortunately, I've been running into nothing but issues. Some I've worked around, but this last one is a blocker for me. What I'm attempting to do is share some data between steps. I've seen the Gherkin tests for Pester Gherkin on this, and you'd think it would be super simple and a non-issue, but I'm not having that luck.

When I first implemented the tests locally I found that I couldn't share data at all without first declaring a variable with script scope in BeforeEachScenario. I made things easy on myself and declared a single hashtable with $script:context = @{}. Then in my steps I'd access things like $context.moduleName. Everything worked great. Then I added AppVeyor build support and it failed there. With some Write-Host calls sprinkled in a few places I determined $context was $null, despite initializing in BeforeEachScenario (and I proved that was called as expected). Thinking maybe I'd need to qualify the scope everywhere and not just when initializing it I changed all the steps to use $script:context, but that caused things to also fail locally. I'm at a loss as to what the problem is here.

The code in question can be found in my [GitHub repo](https://github.com/wekempf/PoshForge/blob/master/tests/steps/PoshForge.steps.ps1).

## 2. Describe Your Environment

Pester version     : 3.4.0 C:\Program Files\WindowsPowerShell\Modules\Pester\3.4.0\Pester.psd1
PowerShell version : 5.1.17134.112
OS version         : Microsoft Windows NT 10.0.17134.0

Not entirely sure about the AppVeyor environment.

## 3. Expected Behavior

I expect to be able to use variables across steps.

## 4.Current Behavior

A variable initialized in BeforeEachScenario is then $null in steps.

## 5. Possible Solution

???

## 6. Context

I'm trying to use Gherkin tests within an AppVeyor build.

**Comment by it-praktyk** [comment]
What about using global variables?

**Comment by Jaykul** [comment]
I agree you should not need global variables ;-)

**Comment by nohwnd** [comment]
I saw that gherkin is 1) making it's own module for every scenario to give it clean scope, and also it is invoking scripts in "weird" scopes, #1162 I also wrote some instrumenation to debug scope (session state) related issues, it's described in the linked issue. Hopefully it will help you debug it.

**Comment by fourpastmidnight** [comment]
@nohwnd Note that this issue is exactly the reason why I use `$Script:Context` in all of my step definition script blocks--so that the variables I declare are definitely available in future blocks, because I ran across instances where they were not (though, I don't remember the particulars since I have found this method to "work"). I do like the mention of using a `BeforeEachScenario` when I came across this issue a few weeks ago looking through the open issues--I just haven't had a chance to update my code. But, it would be nice to not have to do this at all. So hopefull, if we can resolve all the issues related to #1162, including this one, this would be a huge achievement.
